qcodes/interactive_widget.py

```diff
@@ -0,0 +1,436 @@
+    max_nrows: int = 30,
+) -> GridspecLayout:
+    """Generates a `GridspecLayout` of the ``nested_keys`` in ``table`` which is
+    put inside of ``box``.
+
+    Whenever the table has less than ``max_nrows`` rows, the table is
+    displayed in 3 columns, otherwise it's 2 columns.
+    """
+
+    def _should_expand(x: Any) -> bool:
+        return isinstance(x, dict) and x != {}
+
+    col_widths = [8, 16, 30]
+    selected_table = _get_in(nested_keys, table)
+    nrows = (
+        sum(
+            len(v) if _should_expand(v) else 1 for v in selected_table.values()
+        )
+        + 1
+    )
+    ncols = 3
+
+    if nrows > max_nrows:
+        nrows = len(selected_table) + 1
+        col_widths.pop(1)
+        ncols = 2
+
+    grid = GridspecLayout(nrows, col_widths[-1])
+    update = partial(_update_nested_dict_browser, table=table, box=box)
+
+    # Header
+    title = " ► ".join(nested_keys)
+    grid[0, :-1] = button(title, "success")
+    up_click = update(nested_keys[:-1])
+    grid[0, -1] = button("↰", "info", up_click)
+
+    # Body
+
+    i = 1
+    for k, v in selected_table.items():
+        row_length = len(v) if _should_expand(v) and ncols == 3 else 1
+        but = button(k, "info", up_click)
+        grid[i : i + row_length, : col_widths[0]] = but
+        if _should_expand(v):
+            if ncols == 3:
+                for k_, v_ in v.items():
+                    but = button(k_, "danger", update([*nested_keys, k]))
+                    grid[i, col_widths[0] : col_widths[1]] = but
+                    if _should_expand(v_):
+                        sub_keys = ", ".join(v_.keys())
+                        but = button(
+                            sub_keys, "warning", update([*nested_keys, k, k_])
+                        )
+                    else:
+                        but = label(str(v_))
+                    grid[i, col_widths[1] :] = but
+                    i += 1
+            else:
+                sub_keys = ", ".join(v.keys())
+                grid[i, col_widths[0] :] = button(
+                    sub_keys, "danger", update([*nested_keys, k])
+                )
+                i += 1
+        else:
+            grid[i, col_widths[0] :] = label(str(v))
+            i += 1
+    return grid
+
+
+def nested_dict_browser(
+    nested_dict: Dict[Any, Any], nested_keys: Sequence[str] = ()
+) -> Box:
+    """Returns a widget to interactive browse a nested dictionary."""
+    box = Box([])
+    _update_nested_dict_browser(nested_keys, nested_dict, box)(None)
+    return box
+
+
+def _plot_ds(ds: DataSet) -> None:
+    try:
+        # `get_data_by_id` might fail
+        nplots = len(get_data_by_id(ds.run_id))  # TODO: might be a better way
+        nrows = math.ceil(nplots / 2) if nplots != 1 else 1
+        ncols = 2 if nplots != 1 else 1
+        fig, axes = plt.subplots(nrows, ncols, figsize=(6 * ncols, 4 * nrows))
+        # `plot_dataset` might also fail.
+        plot_dataset(ds, axes=axes.flatten())
```

Comment:
`plot_dataset` has `**kwargs` which are passed internally to `plt.subplots`, so i suggest to:
* leave the rows/cols logic to `plot_dataset`, OR apply it perhaps after the `plot_dataset` returns the axes if really needed?
* pass the figure created here to `plot_dataset` explicitly - when i played with the widget i somehow managed to screw the place where the plots should appear, so they started showing up not in the tab but elsewhere. So i guess the remedy is to create a new explicit figure her eand pass it to `plot_dataset` (if that somehow doesn't work, let's improve `plot_dataset` to fix that)

Reply:
In 49e4147e5, I let plot_dataset figure out the fig sizes and axes.